File: java/src/main/java/com/logicalclocks/hsfs/metadata/CodeApi.java
/*
 * Copyright (c) 2021 Logical Clocks AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.logicalclocks.hsfs.metadata;

import com.damnhandy.uri.template.UriTemplate;
import com.logicalclocks.hsfs.EntityEndpointType;
import com.logicalclocks.hsfs.FeatureStoreException;
import com.logicalclocks.hsfs.TrainingDataset;
import lombok.NonNull;
import org.apache.http.client.methods.HttpPost;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

import static com.logicalclocks.hsfs.metadata.HopsworksClient.PROJECT_PATH;
import static com.logicalclocks.hsfs.metadata.HopsworksClient.getInstance;

public class CodeApi {

  public static final String ENTITY_ROOT_PATH = "{/dataSetType}";
  public static final String ENTITY_ID_PATH = ENTITY_ROOT_PATH + "{/dataSetId}";
  public static final String CODE_PATH = ENTITY_ID_PATH + "/code{?entityId,type,databricksClusterId}";

  private static final Logger LOGGER = LoggerFactory.getLogger(CodeApi.class);

  private EntityEndpointType entityType;

  public CodeApi(@NonNull EntityEndpointType entityType) {
    this.entityType = entityType;
  }

  public void post(FeatureGroupBase featureGroup, Code code, String entityId, Code.RunType type, String browserHostName)
          throws FeatureStoreException, IOException {
    post(featureGroup.getFeatureStore().getProjectId(), featureGroup.getFeatureStore().getId(),
            featureGroup.getId(), code, entityId, type, browserHostName);
  }

  public void post(TrainingDataset trainingDataset, Code code, String entityId, Code.RunType type,
                   String browserHostName)
          throws FeatureStoreException, IOException {
    post(trainingDataset.getFeatureStore().getProjectId(), trainingDataset.getFeatureStore().getId(),
            trainingDataset.getId(), code, entityId, type, browserHostName);
  }

  private void post(Integer projectId, Integer featureStoreId, Integer dataSetId, Code code,
                    String entityId, Code.RunType type, String browserHostName)
          throws FeatureStoreException, IOException {
    HopsworksClient hopsworksClient = getInstance();
    String pathTemplate = PROJECT_PATH + FeatureStoreApi.FEATURE_STORE_PATH + CODE_PATH;

    String uri = UriTemplate.fromTemplate(pathTemplate)
            .set("projectId", projectId)
            .set("fsId", featureStoreId)
            .set("dataSetType", entityType.getValue())
            .set("dataSetId", dataSetId)
            .set("entityId", entityId)
            .set("type", type)
            .set("databricksClusterId", browserHostName)
            .expand();

    LOGGER.info("Sending metadata request: " + uri);
    HttpPost postRequest = new HttpPost(uri);
    postRequest.setEntity(hopsworksClient.buildStringEntity(code));
    hopsworksClient.handleRequest(postRequest);
  }
}

File: python/hsfs/core/code_api.py
#
#   Copyright 2021 Logical Clocks AB
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#

from hsfs import client


class CodeApi:
    def __init__(self, feature_store_id, entity_type):
        """Code endpoint for `trainingdatasets` and `featuregroups` resource.

        :param feature_store_id: id of the respective featurestore
        :type feature_store_id: int
        :param entity_type: "trainingdatasets" or "featuregroups"
        :type entity_type: str
        """
        self._feature_store_id = feature_store_id
        self._entity_type = entity_type

    def post(
        self,
        metadata_instance,
        code,
        entity_id,
        code_type,
        databricks_cluster_id=None,
    ):
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
            self._entity_type,
            metadata_instance.id,
            "code",
        ]

        headers = {"content-type": "application/json"}

        query_params = {
            "entityId": entity_id,
            "type": code_type,
            "databricksClusterId": databricks_cluster_id,
        }

        _client._send_request(
            "POST", path_params, query_params, headers=headers, data=code.json()
        )
